L1Trigger/L1TTrackerDTC/src/DTC.cc

```diff
@@ -0,0 +1,145 @@
+  void DTC::consume(const vector<TTStubRef>& ttStubRefStream, const int& channelId) {
+    for (const TTStubRef& ttStubRef : ttStubRefStream)
+      stubs_.push_back(move(new Stub(settings_, ttStubRef, modules_[channelId])));  // convert TTStub
+  }
+
+  // board level routing in two steps and product filling
+  void DTC::produce(TTDTC& product) {
+    Stubsss moduleStubs(settings_->numRoutingBlocks(),
+                        Stubss(settings_->numModulesPerRoutingBlock()));  // empty input container
+    Stubss blockStubs(settings_->numRoutingBlocks());                     // empty intermediate container
+    Stubss regionStubs(settings_->numOverlappingRegions());               // empty output container
+
+    // fill input
+    for (Stub* stub : stubs_)
+      if (stub->valid())  // pt and eta cut
+        moduleStubs[stub->blockId()][stub->channelId()].push_back(stub);
+
+    // sort stubs by bend
+    for (auto& block : moduleStubs)
+      for (auto& channel : block)
+        sort(channel.begin(), channel.end(), [](Stub* lhs, Stub* rhs) { return abs(lhs->bend()) < abs(rhs->bend()); });
+
+    // router step 1: merges stubs of all modules connected to one routing block into one stream
+    for (int routingBlock = 0; routingBlock < settings_->numRoutingBlocks(); routingBlock++)
+      merge(moduleStubs[routingBlock], blockStubs[routingBlock]);
+
+    // router step 2: merges stubs of all routing blocks and splits stubs into one stream per overlapping region
+    split(blockStubs, regionStubs);
+
+    // fill product
+    for (int channel = 0; channel < settings_->numOverlappingRegions(); channel++) {
+      Stubs& stubs = regionStubs[channel];
+      TTDTC::Stream& stream = product.dtcStream(region_, board_, channel);
+
+      if (settings_->enableTruncation())  // truncate if desired
+        stubs.resize(min((int)stubs.size(), settings_->maxFramesChannelOutput()));
+
+      // remove all gaps between end and last stub
+      for (auto it = stubs.end(); it != stubs.begin();)
+        it = (*--it) ? stubs.begin() : stubs.erase(it);
+
+      stream.reserve(stubs.size());
+      for (const Stub* stub : stubs) {
+        if (stub)
+          stream.emplace_back(stub->ttStubRef(), stub->frame(channel));
+        else  // use default constructed TTDTC::Pair to represent gaps
+          stream.emplace_back();
+      }
+    }
+  }
+
+  // router step 1: merges stubs of all modules connected to one routing block into one stream
+  void DTC::merge(Stubss& inputs, Stubs& output) {
+    Stubss stacks(inputs.size());  // for each input one fifo
+
+    // clock accurate firmware emulation, each while trip describes one clock tick
+    while (!all_of(inputs.begin(), inputs.end(), [](const Stubs& channel) { return channel.empty(); }) or
+           !all_of(stacks.begin(), stacks.end(), [](const Stubs& channel) {
+             return channel.empty();
+           })) {  // work as long as needed to merge all stubs
```

Comment:
again recommend putting comments on separate lines to avoid awkward breaks

Reply:
done